modules/modelSetup/mixin/ModelSetupDiffusionLossMixin.py

```diff
@@ -171,6 +172,39 @@ def __min_snr_weight(
             snr += 1
         snr_weight = torch.div(min_snr_gamma, snr).float().to(device)
         return snr_weight
+    
+    def __debiased_estimation_weight(
+        self,
+        timesteps: Tensor,
+        device: torch.device
+    ) -> Tensor:
+        if self.__coefficients:
+            all_snr = (self.__coefficients.sqrt_alphas_cumprod /
+                       self.__coefficients.sqrt_one_minus_alphas_cumprod) ** 2
+            all_snr.to(device)
+            snr = all_snr[timesteps]
+        else:
+            alphas_cumprod = self.__alphas_cumprod_fun(timesteps, 1)
+            snr = ((alphas_cumprod ** 0.5) / (1.0 - alphas_cumprod) ** 0.5) ** 2
+        
+        weight = snr
+        torch.minimum(weight, torch.full_like(snr, 1000), out=weight)
+        torch.sqrt(weight, out=weight)
+        torch.inverse(weight, out=weight)
```

Comment:
Better test this code path with zero-terminal SNR enabled. Can't this result in a division by zero in that case?

Reply:
Might want to pull out the pen and paper and derive what this weighting should look like when predicting velocity instead of epsilon.

Reply:
I changed it to be `torch.clamp` instead.

Reply:
I discussed this with Nero in Discord, seems like it doesn't need any adjustments.

Reply:
Don't think this is resolved (see Discord). Just like minimum-SNR gamma does, under v-prediction we should use SNR+1 instead of SNR (so, weighting of 1/sqrt(SNR+1) instead of 1/sqrt(SNR)).

Reasoning is the same as min-SNR gamma reasoning: eps-prediction is implicitly weighted by SNR, so they divide through by SNR. Under v-prediction, the algebra for that comes out to be SNR+1.

So why 1/sqrt(SNR+1) instead of 1/sqrt(SNR)? Well the paper this PR is based on originally goes on about wanting to use the 1/SNR weight (which would be x0 prediction), decided it didn't work well, and decided 1/sqrt(SNR) worked better. There's no real mathematical justification for that. The paper did not consider the case of v-prediction, so likewise there's no real mathematical justification for 1/sqrt(SNR+1), but it's at least congruent with how the paper transforms the 1/SNR weighting to 1/sqrt(SNR) and what that would look like under v-prediction.

Reply:
I've done `SNR += 1` for both P2 and debiased.

Reply:
Resolving this, the relevant open discussion for P2 is here: https://github.com/Nerogar/OneTrainer/pull/212#discussion_r1536864469